src/gp_engine.cpp

```diff
@@ -3,54 +3,171 @@
+    rescaling_counts_.conservativeResize(GetAllocatedPLVCount());
+  }
+  // Initialize new work space.
+  Assert(plvs_.back().rows() == MmappedNucleotidePLV::base_count_ &&
+             plvs_.back().cols() ==
+                 static_cast<Eigen::Index>(site_pattern_.PatternCount()),
+         "Didn't get the right shape of PLVs out of Subdivide.");
+  for (size_t i = old_plv_count; i < GetPaddedPLVCount(); i++) {
+    rescaling_counts_[i] = 0;
+    plvs_.at(i).setZero();
+  }
+  // Reindex work space to match DAG.
+  if (node_reindexer.has_value()) {
+    ReindexPLVs(node_reindexer.value(), old_node_count);
+  }
+}
+
+void GPEngine::GrowGPCSPs(const size_t new_gpcsp_count,
+                          std::optional<Reindexer> gpcsp_reindexer) {
+  size_t old_gpcsp_count = gpcsp_count_;
+  gpcsp_count_ = new_gpcsp_count;
+  // Reallocate more space if needed.
+  if (GetPaddedGPCSPCount() > gpcsp_alloc_) {
+    gpcsp_alloc_ = size_t(ceil(double(GetPaddedGPCSPCount()) * gpcsp_resizing_factor_));
+    branch_lengths_.conservativeResize(gpcsp_alloc_);
+    hybrid_marginal_log_likelihoods_.conservativeResize(gpcsp_alloc_);
+    log_likelihoods_.conservativeResize(gpcsp_alloc_, site_pattern_.PatternCount());
+  }
+  // Initialize new work space.
+  for (size_t i = old_gpcsp_count; i < GetPaddedGPCSPCount(); i++) {
+    branch_lengths_[i] = default_branch_length_;
+    hybrid_marginal_log_likelihoods_[i] = DOUBLE_NEG_INF;
+  }
+  // Reindex work space to match DAG.
+  if (gpcsp_reindexer.has_value()) {
+    ReindexGPCSPs(gpcsp_reindexer.value(), old_gpcsp_count);
+  }
+}
+
+void GPEngine::ReindexPLVs(const Reindexer node_reindexer,
+                           const size_t old_node_count) {
+  Assert(node_reindexer.size() == node_count_,
+         "Node Reindexer is the wrong size for GPEngine.");
+  Assert(node_reindexer.IsValid(node_count_), "Node Reindexer is not valid.");
+
+  // Expand node_reindexer into plv_reindexer.
+  Reindexer plv_reindexer = Reindexer(node_count_ * plv_count_per_node_);
+  DoubleVector plv_test_in = DoubleVector(node_count_, 0.);
+  DoubleVector plv_test_out = DoubleVector(node_count_, 0.);
+  size_t new_idx = old_node_count * plv_count_per_node_;
+  for (size_t i = 0; i < node_count_; i++) {
+    const size_t old_node = i;
+    const size_t new_node = node_reindexer.GetNewIndexByOldIndex(i);
+    for (size_t j = 0; j < plv_count_per_node_; j++) {
+      // Either get input plv_index from old plvs, or take next new plv_index (data is
+      // irrelevant).
+      size_t old_plv = 0;
+      if (old_node < old_node_count) {
+        old_plv = PLVHandler::GetPLVIndex(j, old_node_count, old_node);
+      } else {
+        old_plv = new_idx;
+        new_idx++;
+      }
+      const size_t new_plv = PLVHandler::GetPLVIndex(j, node_count_, new_node);
+      plv_reindexer.SetReindex(old_plv, new_plv);
+    }
+  }
+  Assert(plv_reindexer.IsValid(GetPLVCount()), "PLV Reindexer is not valid.");
+
+  // Reindex PLVs in place.
```

Comment:
The details of how this works are a little obscure to me. I'm not sure if pulling it out as a separate function would help, or some comments?